File: native/android/AetherBand/app/src/main/java/app/aether/band/HeartRateCentral.kt
package app.aether.band

import android.Manifest
import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Base64
import androidx.core.content.ContextCompat
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

data class ScanHit(val id: String, val name: String, val rssi: Int)

class NativeBleError(val jsName: String, override val message: String) : Exception(message)

class HeartRateCentral(private val context: Context) {
  var onHits: ((List<ScanHit>) -> Unit)? = null
  var onShowPicker: (() -> Unit)? = null
  var onHidePicker: (() -> Unit)? = null
  var onNotify: ((String, ByteArray, Long) -> Unit)? = null
  var onDisconnect: (() -> Unit)? = null
  var onFlush: ((List<Map<String, String>>) -> Unit)? = null
  var onKeepAlive: ((Boolean) -> Unit)? = null

  private val handler = Handler(Looper.getMainLooper())
  private val adapter: BluetoothAdapter? =
    (context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager).adapter
  private val prefs = context.getSharedPreferences("aether.ble", Context.MODE_PRIVATE)
  private val seen = ConcurrentHashMap<String, BluetoothDevice>()
  private val hits = mutableListOf<ScanHit>()
  private val characteristics = ConcurrentHashMap<String, BluetoothGattCharacteristic>()
  private val services = mutableSetOf<String>()
  private val buffer = mutableListOf<Map<String, String>>()
  private var lastBufferAt = 0L
  private var gatt: BluetoothGatt? = null
  private var scanAll = false
  private var reconnecting = false
  private var pickResume: ((Result<Map<String, String>>) -> Unit)? = null
  private var connectResume: ((Result<Unit>) -> Unit)? = null
  private var readResume: ((Result<String>) -> Unit)? = null
  private var scanning = false

  private var rememberedId: String?
    get() = prefs.getString("peripheral", null)
    set(value) { prefs.edit().putString("peripheral", value).apply() }

  private var rememberedName: String
    get() = prefs.getString("name", "HR strap") ?: "HR strap"
    set(value) { prefs.edit().putString("name", value).apply() }

  fun rememberedDevices(): List<Map<String, String>> {
    val id = rememberedId ?: return emptyList()
    return listOf(mapOf("id" to id, "name" to rememberedName))
  }

  fun enableReconnect() {
    reconnecting = true
  }

  fun flushBuffer() {
    if (buffer.isEmpty()) return
    val copy = buffer.toList()
    buffer.clear()
    onFlush?.invoke(copy)
  }

  @SuppressLint("MissingPermission")
  fun cancelPicker() {
    stopScan()
    onHidePicker?.invoke()
    pickResume?.invoke(Result.failure(NativeBleError("NotFoundError", "Pairing cancelled.")))
    pickResume = null
  }

  fun cancelIfPicking() {
    if (pickResume != null) cancelPicker()
  }

  @SuppressLint("MissingPermission")
  fun pick(hit: ScanHit) {
    stopScan()
    onHidePicker?.invoke()
    rememberedId = hit.id
    rememberedName = hit.name
    pickResume?.invoke(Result.success(mapOf("id" to hit.id, "name" to hit.name)))
    pickResume = null
  }

  fun currentHits(): List<ScanHit> = hits.toList()

  @SuppressLint("MissingPermission")
  fun requestDevice(scanAll: Boolean, done: (Result<Map<String, String>>) -> Unit) {
    if (!hasPermission()) {
      done(Result.failure(NativeBleError("SecurityError", "Allow Bluetooth for Aether in Settings.")))
      return
    }
    val adapter = adapter
    if (adapter == null || !adapter.isEnabled) {
      done(Result.failure(NativeBleError("SecurityError", "Turn on Bluetooth on this phone.")))
      return
    }
    this.scanAll = scanAll
    hits.clear()
    onHits?.invoke(emptyList())
    onShowPicker?.invoke()
    pickResume = done
    val settings = ScanSettings.Builder().setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY).build()
    stopScan()
    adapter.bluetoothLeScanner?.startScan(null, settings, scanCallback)
    scanning = true
  }

  @SuppressLint("MissingPermission")
  fun connect(deviceId: String, done: (Result<Unit>) -> Unit) {
    if (!hasPermission()) {
      done(Result.failure(NativeBleError("SecurityError", "Allow Bluetooth for Aether in Settings.")))
      return
    }
    val adapter = adapter ?: run {
      done(Result.failure(NativeBleError("NotFoundError", "No heart-rate strap found.")))
      return
    }
    val device = seen[deviceId] ?: try {
      adapter.getRemoteDevice(deviceId)
    } catch (_: IllegalArgumentException) {
      done(Result.failure(NativeBleError("NotFoundError", "No heart-rate strap found.")))
      return
    }
    characteristics.clear()
    services.clear()
    connectResume = done
    gatt?.close()
    gatt = device.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
  }

  @SuppressLint("MissingPermission")
  fun disconnect() {
    reconnecting = false
    onKeepAlive?.invoke(false)
    gatt?.disconnect()
    gatt?.close()
    gatt = null
  }

  fun requireService(uuid: String) {
    if (!services.contains(canonical(uuid))) {
      throw NativeBleError("NetworkError", "That device did not expose $uuid.")
    }
  }

  fun requireCharacteristic(uuid: String) {
    if (characteristic(uuid) == null) {
      throw NativeBleError("NetworkError", "Missing characteristic $uuid.")
    }
  }

  @SuppressLint("MissingPermission")
  fun startNotifications(uuid: String) {
    val gatt = gatt
    val char = characteristic(uuid) ?: throw NativeBleError("NetworkError", "Missing characteristic $uuid.")
    if (gatt == null) throw NativeBleError("NetworkError", "The strap dropped Bluetooth.")
    gatt.setCharacteristicNotification(char, true)
    val cccd = char.getDescriptor(CCCD) ?: return
    if (Build.VERSION.SDK_INT >= 33) {
      gatt.writeDescriptor(cccd, BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE)
    } else {
      @Suppress("DEPRECATION")
      cccd.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
      @Suppress("DEPRECATION")
      gatt.writeDescriptor(cccd)
    }
  }

  @SuppressLint("MissingPermission")
  fun stopNotifications(uuid: String) {
    val gatt = gatt ?: return
    val char = characteristic(uuid) ?: return
    gatt.setCharacteristicNotification(char, false)
  }

  @SuppressLint("MissingPermission")
  fun readValue(uuid: String, done: (Result<String>) -> Unit) {
    val gatt = gatt
    val char = characteristic(uuid)
    if (gatt == null || char == null) {
      done(Result.failure(NativeBleError("NetworkError", "Missing characteristic $uuid.")))
      return
    }
    readResume = done
    gatt.readCharacteristic(char)
  }

  private fun characteristic(uuid: String): BluetoothGattCharacteristic? =
    characteristics[canonical(uuid)]

  private fun hasPermission(): Boolean {
    val connect = if (Build.VERSION.SDK_INT >= 31) {
      ContextCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_CONNECT) ==
        PackageManager.PERMISSION_GRANTED &&
        ContextCompat.checkSelfPermission(context, Manifest.permission.BLUETOOTH_SCAN) ==
        PackageManager.PERMISSION_GRANTED
    } else {
      ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) ==
        PackageManager.PERMISSION_GRANTED
    }
    return connect
  }

  @SuppressLint("MissingPermission")
  private fun stopScan() {
    if (!scanning) return
    scanning = false
    try {
      adapter?.bluetoothLeScanner?.stopScan(scanCallback)
    } catch (_: Exception) {
      /* already stopped */
    }
  }

  private val scanCallback = object : ScanCallback() {
    override fun onScanFailed(errorCode: Int) {
      handler.post {
        scanning = false
        onHidePicker?.invoke()
        pickResume?.invoke(Result.failure(NativeBleError("NetworkError", "Bluetooth failed.")))
        pickResume = null
      }
    }

    @SuppressLint("MissingPermission")
    override fun onScanResult(callbackType: Int, result: ScanResult) {
      val device = result.device
      val address = device.address ?: return
      seen[address] = device
      val name = result.scanRecord?.deviceName ?: device.name ?: "Heart rate strap"
      if (!matchesFilter(name)) return
      val hit = ScanHit(address, name, result.rssi)
      handler.post {
        val idx = hits.indexOfFirst { it.id == address }
        if (idx >= 0) hits[idx] = hit else hits.add(hit)
        hits.sortByDescending { it.rssi }
        onHits?.invoke(hits.toList())
      }
    }
  }

  private val gattCallback = object : BluetoothGattCallback() {
    @SuppressLint("MissingPermission")
    override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
      handler.post {
        if (newState == BluetoothProfile.STATE_CONNECTED) {
          rememberedId = gatt.device.address
          rememberedName = gatt.device.name?.ifBlank { null } ?: rememberedName
          gatt.discoverServices()
          onKeepAlive?.invoke(true)
        } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
          connectResume?.invoke(Result.failure(NativeBleError("NetworkError", "The strap dropped Bluetooth.")))
          connectResume = null
          onDisconnect?.invoke()
          if (reconnecting) {
            handler.postDelayed({
              val id = rememberedId ?: return@postDelayed
              connect(id) { }
            }, 1500)
          } else {
            onKeepAlive?.invoke(false)
          }
        }
      }
    }

    override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
      handler.post {
        if (status != BluetoothGatt.GATT_SUCCESS) {
          connectResume?.invoke(Result.failure(NativeBleError("NetworkError", "The strap dropped Bluetooth.")))
          connectResume = null
          return@post
        }
        for (service in gatt.services) {
          services.add(canonical(service.uuid.toString()))
          for (char in service.characteristics) {
            characteristics[canonical(char.uuid.toString())] = char
          }
        }
        connectResume?.invoke(Result.success(Unit))
        connectResume = null
      }
    }

    @Deprecated("Deprecated in Java")
    override fun onCharacteristicRead(
      gatt: BluetoothGatt,
      characteristic: BluetoothGattCharacteristic,
      status: Int,
    ) {
      val value = characteristic.value ?: ByteArray(0)
      onCharacteristicRead(gatt, characteristic, value, status)
    }

    override fun onCharacteristicRead(
      gatt: BluetoothGatt,
      characteristic: BluetoothGattCharacteristic,
      value: ByteArray,
      status: Int,
    ) {
      handler.post {
        val b64 = Base64.encodeToString(value, Base64.NO_WRAP)
        if (status != BluetoothGatt.GATT_SUCCESS) {
          readResume?.invoke(Result.failure(NativeBleError("NetworkError", "Bluetooth failed.")))
        } else {
          readResume?.invoke(Result.success(b64))
        }
        readResume = null
      }
    }

    @Deprecated("Deprecated in Java")
    override fun onCharacteristicChanged(
      gatt: BluetoothGatt,
      characteristic: BluetoothGattCharacteristic,
    ) {
      onCharacteristicChanged(gatt, characteristic, characteristic.value ?: ByteArray(0))
    }

    override fun onCharacteristicChanged(
      gatt: BluetoothGatt,
      characteristic: BluetoothGattCharacteristic,
      value: ByteArray,
    ) {
      val key = canonical(characteristic.uuid.toString())
      val now = System.currentTimeMillis()
      handler.post {
        onNotify?.invoke(key, value, now)
        if (now - lastBufferAt >= 20_000) {
          lastBufferAt = now
          val b64 = Base64.encodeToString(value, Base64.NO_WRAP)
          buffer.add(mapOf("uuid" to key, "data" to b64, "t" to now.toString()))
          if (buffer.size > 2600) {
            buffer.subList(0, buffer.size - 2600).clear()
          }
        }
      }
    }
  }

  private fun matchesFilter(name: String): Boolean {
    if (scanAll) return true
    val lower = name.lowercase()
    return PREFIXES.any { lower.contains(it) }
  }

  companion object {
    private val CCCD: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
    private val PREFIXES = listOf("whoop", "polar", "garmin", "wahoo", "tickr", "hrm", "coospo", "magene", "scosche")

    fun canonical(raw: String): String {
      val value = raw.trim().lowercase()
      val aliases = mapOf(
        "heart_rate" to "180d",
        "heart_rate_measurement" to "2a37",
        "battery_service" to "180f",
        "battery_level" to "2a19",
        "pulse_oximeter" to "1822",
        "plx_continuous_measurement" to "2a5f",
        "plx_spot_check_measurement" to "2a5e",
        "health_thermometer" to "1809",
        "temperature_measurement" to "2a1c",
        "intermediate_temperature" to "2a1e",
      )
      aliases[value]?.let { return it }
      if (value.length == 36) {
        val hex = value.replace("-", "")
        if (hex.startsWith("0000") && hex.endsWith("00001000800000805f9b34fb")) {
          return hex.substring(4, 8)
        }
      }
      return value
    }
  }
}

File: native/ios/AetherBand/AetherBand/HeartRateCentral.swift
import Combine
import CoreBluetooth
import Foundation

struct BleScanHit: Identifiable, Equatable {
  let id: UUID
  var name: String
  var rssi: Int
}

enum NativeBleError: LocalizedError {
  case cancelled
  case poweredOff
  case unauthorized
  case notFound
  case disconnected
  case missingService(String)
  case missingCharacteristic(String)

  var name: String {
    switch self {
    case .cancelled, .notFound: return "NotFoundError"
    case .poweredOff, .unauthorized: return "SecurityError"
    case .disconnected: return "NetworkError"
    case .missingService, .missingCharacteristic: return "NetworkError"
    }
  }

  var errorDescription: String? {
    switch self {
    case .cancelled: return "Pairing cancelled."
    case .poweredOff: return "Turn on Bluetooth on this iPhone."
    case .unauthorized: return "Allow Bluetooth for Aether in Settings."
    case .notFound: return "No heart-rate strap found."
    case .disconnected: return "The strap dropped Bluetooth."
    case .missingService(let uuid):
      return "That device did not expose \(uuid)."
    case .missingCharacteristic(let uuid):
      return "Missing characteristic \(uuid)."
    }
  }
}

final class HeartRateCentral: NSObject, ObservableObject {
  @Published var hits: [BleScanHit] = []
  @Published var showPicker = false
  @Published var statusText = "Ready"

  private var central: CBCentralManager!
  private var peripheral: CBPeripheral?
  private var characteristics: [String: CBCharacteristic] = [:]
  private var services: Set<String> = []
  private var scanAll = false
  private var pickContinuation: CheckedContinuation<[String: String], Error>?
  private var connectContinuation: CheckedContinuation<Void, Error>?
  private var readWaiters: [String: CheckedContinuation<String, Error>] = [:]
  private var readyWaiters: [CheckedContinuation<Void, Error>] = []
  private var rememberedId: UUID? {
    get {
      UserDefaults.standard.string(forKey: "aether.ble.peripheral").flatMap(UUID.init(uuidString:))
    }
    set {
      UserDefaults.standard.set(newValue?.uuidString, forKey: "aether.ble.peripheral")
    }
  }
  private var rememberedName: String {
    get { UserDefaults.standard.string(forKey: "aether.ble.name") ?? "HR strap" }
    set { UserDefaults.standard.set(newValue, forKey: "aether.ble.name") }
  }
  private var seen: [UUID: CBPeripheral] = [:]
  private var reconnecting = false
  var onNotify: ((String, Data) -> Void)?
  var onDisconnect: (() -> Void)?
  var onFlush: (([[String: String]]) -> Void)?
  private var buffer: [[String: String]] = []
  private var lastBufferAt: TimeInterval = 0

  override init() {
    super.init()
    central = CBCentralManager(delegate: self, queue: .main)
  }

  func waitUntilReady() async throws {
    switch central.state {
    case .poweredOn:
      return
    case .poweredOff:
      throw NativeBleError.poweredOff
    case .unauthorized:
      throw NativeBleError.unauthorized
    default:
      try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
        readyWaiters.append(cont)
      }
    }
  }

  func requestDevice(scanAll: Bool) async throws -> [String: String] {
    try await waitUntilReady()
    self.scanAll = scanAll
    hits = []
    showPicker = true
    statusText = "Scanning…"
    central.scanForPeripherals(withServices: nil, options: [CBCentralManagerScanOptionAllowDuplicatesKey: true])
    return try await withCheckedThrowingContinuation { cont in
      pickContinuation = cont
    }
  }

  func cancelIfPicking() {
    if pickContinuation != nil {
      cancelPicker()
    }
  }

  func cancelPicker() {
    central.stopScan()
    showPicker = false
    if let pickContinuation {
      self.pickContinuation = nil
      pickContinuation.resume(throwing: NativeBleError.cancelled)
    }
  }

  func pick(_ hit: BleScanHit) {
    central.stopScan()
    showPicker = false
    rememberedId = hit.id
    rememberedName = hit.name
    let info = ["id": hit.id.uuidString, "name": hit.name]
    pickContinuation?.resume(returning: info)
    pickContinuation = nil
  }

  func rememberedDevices() -> [[String: String]] {
    guard let rememberedId else { return [] }
    return [["id": rememberedId.uuidString, "name": rememberedName]]
  }

  func connect(deviceId: String) async throws {
    try await waitUntilReady()
    guard let uuid = UUID(uuidString: deviceId) else { throw NativeBleError.notFound }
    let known = seen[uuid] ?? central.retrievePeripherals(withIdentifiers: [uuid]).first
    guard let target = known else {
      throw NativeBleError.notFound
    }
    characteristics = [:]
    services = []
    peripheral = target
    target.delegate = self
    if target.state == .connected {
      target.discoverServices(nil)
    } else {
      central.connect(target, options: nil)
    }
    try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
      connectContinuation = cont
    }
  }

  func disconnect() {
    reconnecting = false
    if let peripheral {
      central.cancelPeripheralConnection(peripheral)
    }
  }

  func requireService(_ uuid: String) throws {
    let key = Self.canonical(uuid)
    if !services.contains(key) {
      throw NativeBleError.missingService(uuid)
    }
  }

  func requireCharacteristic(_ uuid: String) throws {
    if characteristic(uuid) == nil {
      throw NativeBleError.missingCharacteristic(uuid)
    }
  }

  func startNotifications(_ uuid: String) throws {
    guard let peripheral, let char = characteristic(uuid) else {
      throw NativeBleError.missingCharacteristic(uuid)
    }
    peripheral.setNotifyValue(true, for: char)
  }

  func stopNotifications(_ uuid: String) {
    guard let peripheral, let char = characteristic(uuid) else { return }
    peripheral.setNotifyValue(false, for: char)
  }

  func readValue(_ uuid: String) async throws -> String {
    guard let peripheral, let char = characteristic(uuid) else {
      throw NativeBleError.missingCharacteristic(uuid)
    }
    return try await withCheckedThrowingContinuation { cont in
      readWaiters[Self.canonical(uuid)] = cont
      peripheral.readValue(for: char)
    }
  }

  func flushBuffer() {
    guard !buffer.isEmpty else { return }
    let copy = buffer
    buffer.removeAll()
    onFlush?(copy)
  }

  func enableReconnect() {
    reconnecting = true
  }

  private func characteristic(_ uuid: String) -> CBCharacteristic? {
    characteristics[Self.canonical(uuid)]
  }

  private func finishConnect() {
    connectContinuation?.resume()
    connectContinuation = nil
    statusText = "Connected"
  }

  private func failConnect(_ error: Error) {
    connectContinuation?.resume(throwing: error)
    connectContinuation = nil
  }

  static func canonical(_ raw: String) -> String {
    let value = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    let aliases: [String: String] = [
      "heart_rate": "180d",
      "heart_rate_measurement": "2a37",
      "battery_service": "180f",
      "battery_level": "2a19",
      "pulse_oximeter": "1822",
      "plx_continuous_measurement": "2a5f",
      "plx_spot_check_measurement": "2a5e",
      "health_thermometer": "1809",
      "temperature_measurement": "2a1c",
      "intermediate_temperature": "2a1e",
    ]
    if let mapped = aliases[value] { return mapped }
    if value.count == 36 {
      let hex = value.split(separator: "-").joined()
      if hex.hasPrefix("0000") && hex.hasSuffix("00001000800000805f9b34fb") {
        return String(hex.dropFirst(4).prefix(4))
      }
    }
    return value
  }

  private func keepName(_ peripheral: CBPeripheral) -> String {
    peripheral.name?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
      ? peripheral.name!
      : "Heart rate strap"
  }

  private func matchesFilter(_ peripheral: CBPeripheral) -> Bool {
    if scanAll { return true }
    let name = (peripheral.name ?? "").lowercased()
    let prefixes = ["whoop", "polar", "garmin", "wahoo", "tickr", "hrm", "coospo", "magene", "scosche"]
    return prefixes.contains { name.contains($0) }
  }
}

extension HeartRateCentral: CBCentralManagerDelegate {
  func centralManagerDidUpdateState(_ central: CBCentralManager) {
    switch central.state {
    case .poweredOn:
      let waiters = readyWaiters
      readyWaiters.removeAll()
      waiters.forEach { $0.resume() }
      if reconnecting, let rememberedId {
        let known = central.retrievePeripherals(withIdentifiers: [rememberedId])
        if let target = known.first {
          peripheral = target
          target.delegate = self
          central.connect(target, options: nil)
        }
      }
    case .poweredOff:
      failReady(NativeBleError.poweredOff)
    case .unauthorized:
      failReady(NativeBleError.unauthorized)
    default:
      break
    }
  }

  private func failReady(_ error: NativeBleError) {
    let waiters = readyWaiters
    readyWaiters.removeAll()
    waiters.forEach { $0.resume(throwing: error) }
    if let pickContinuation {
      self.pickContinuation = nil
      pickContinuation.resume(throwing: error)
    }
  }

  func centralManager(
    _ central: CBCentralManager,
    didDiscover peripheral: CBPeripheral,
    advertisementData: [String: Any],
    rssi RSSI: NSNumber
  ) {
    seen[peripheral.identifier] = peripheral
    guard matchesFilter(peripheral) else { return }
    let name = (advertisementData[CBAdvertisementDataLocalNameKey] as? String)
      ?? peripheral.name
      ?? "Heart rate strap"
    let hit = BleScanHit(id: peripheral.identifier, name: name, rssi: RSSI.intValue)
    if let idx = hits.firstIndex(where: { $0.id == hit.id }) {
      hits[idx] = hit
    } else {
      hits.append(hit)
      hits.sort { $0.rssi > $1.rssi }
    }
  }

  func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
    rememberedId = peripheral.identifier
    rememberedName = keepName(peripheral)
    peripheral.delegate = self
    peripheral.discoverServices(nil)
  }

  func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
    failConnect(error ?? NativeBleError.disconnected)
    if reconnecting {
      DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in
        self?.central.connect(peripheral, options: nil)
      }
    }
  }

  func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
    onDisconnect?()
    if reconnecting {
      statusText = "Reconnecting…"
      DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) { [weak self] in
        self?.central.connect(peripheral, options: nil)
      }
    }
  }
}

extension HeartRateCentral: CBPeripheralDelegate {
  func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
    if let error {
      failConnect(error)
      return
    }
    guard let found = peripheral.services, !found.isEmpty else {
      failConnect(NativeBleError.missingService("heart_rate"))
      return
    }
    for service in found {
      services.insert(Self.canonical(service.uuid.uuidString))
      peripheral.discoverCharacteristics(nil, for: service)
    }
  }

  func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
    service.characteristics?.forEach { char in
      characteristics[Self.canonical(char.uuid.uuidString)] = char
    }
    let pending = peripheral.services?.contains { $0.characteristics == nil } ?? false
    if !pending {
      finishConnect()
    }
  }

  func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
    let key = Self.canonical(characteristic.uuid.uuidString)
    if let error {
      readWaiters[key]?.resume(throwing: error)
      readWaiters[key] = nil
      return
    }
    let data = characteristic.value ?? Data()
    let b64 = data.base64EncodedString()
    if let waiter = readWaiters[key] {
      readWaiters[key] = nil
      waiter.resume(returning: b64)
    }
    onNotify?(key, data)
    let now = Date().timeIntervalSince1970
    if now - lastBufferAt >= 20 {
      lastBufferAt = now
      let ms = String(Int(now * 1000))
      buffer.append(["uuid": key, "data": b64, "t": ms])
      if buffer.count > 2600 { buffer.removeFirst(buffer.count - 2600) }
    }
  }
}
